File: Chapter 09/C#/section_9_3_4_serverless_functions/Handlers/AWSLambdaHandler.cs
using System;
using System.Reactive.Linq;
using Amazon.Lambda.APIGatewayEvents;
using Amazon.Lambda.S3Events;
using Chapter09.ServerlessFunctions.Infrastructure.Aws;

namespace Chapter09.ServerlessFunctions.Handlers;

public class AwsLambdaHandler
{
    /// <summary>
    /// CLOUD 1: AWS LAMBDA (The Imperative Island - Reactive)
    /// 
    /// THE ARCHITECTURAL LESSON: 
    /// Even when using modern Observables, AWS remains "Infrastructure-Heavy." 
    /// Because the platform only hands you metadata, your reactive pipeline 
    /// must contain the "Plumbing" to go fetch the actual data bytes.
    /// </summary>
    /// <param name="eventStream">A stream of proprietary AWS S3 events.</param>
    /// <returns>A stream emitting the AWS-compliant response object.</returns>
    public IObservable<APIGatewayProxyResponse> Handle(IObservable<MockS3Event> eventStream)
    {
        return eventStream.Select(s3Event => 
        {
            /**
             * TEACHING NOTE:
             * Notice the 'Select' operator (the Rx equivalent of Map). We are projecting 
             * AWS-specific metadata into our logic. Because AWS only gives us the 
             * file's "Key", a real implementation would need to chain an SDK call 
             * (e.g. .SelectMany(evt => _s3Client.GetStream$(...))) here. 
             * The "Island" architecture forces the developer to bridge the gap 
             * between "Event" and "Data" manually within the logic.
             */
            string fileName = s3Event.Records[0].S3.Object.Key;
            Console.WriteLine($"      [AWS Lambda] Reactive stream observed: {fileName}");
            Console.WriteLine($"      [AWS Lambda] Plumbing: Projecting SDK fetch into the pipe...");

            return new APIGatewayProxyResponse { 
                StatusCode = 200, 
                Body = "AWS processed reactive stream" 
            };
        });
    }
}

File: Chapter 09/JS/section_9_3_4_serverless_functions/handlers/awsLambdaHandler.js
import { map } from 'rxjs/operators';

/**
 * CLOUD 1: AWS LAMBDA (The Imperative Island - Reactive)
 * * THE ARCHITECTURAL LESSON: 
 * Even in a reactive model, AWS remains "Infrastructure-Heavy." 
 * Because AWS doesn't fetch the file for you, your stream must 
 * contain the plumbing to go get the data.
 * * @param {Observable} event$ - A stream containing the AWS S3 JSON metadata.
 * @param {Object} context - The AWS Lambda context.
 * @returns {Observable} A stream emitting the AWS-compliant response object.
 */
export const handleRequest$ = (event$, context) => {
    return event$.pipe(
        /**
         * TEACHING NOTE:
         * Notice the 'map' operator. We are observing the stream and 
         * projecting the S3 metadata into our logic. But because AWS 
         * only gives us the name (the 'key'), we would have to chain 
         * another observable (like a 'switchMap') to imperatively 
         * call the S3 SDK. The "Island" architecture forces you to 
         * build your own bridge to your data within the pipeline.
         */
        map(event => {
            const fileName = event.Records[0].s3.object.key;
            console.log(`      [AWS Lambda] Stream observed. File: ${fileName}`);
            console.log(`      [AWS Lambda] Plumbing: Projecting SDK fetch into the stream...`);
            
            return {
                statusCode: 200,
                body: JSON.stringify({ message: "AWS processed reactive stream" })
            };
        })
    );
};